cpp/include/raft/neighbors/detail/ivf_pq_build.cuh

```diff
@@ -1858,6 +1861,7 @@ auto build(raft::resources const& handle,
                          trainset.data(),
                          labels.data(),
                          params.kmeans_n_iters,
+                         params.pq_codebook_trainset_fraction,
                          &managed_memory_upstream);
         break;
       case codebook_gen::PER_CLUSTER:
```

Comment:
Since we introduce the trainset-fraction parameter for codebooks, I think it makes sense that this parameter should affect both modes, PER_CLUSTER and PER_SUBSPACE.

Reply:
Make sense. I'll make both modes use the same sampling strategy, either fraction-based or max point-based, after we make a choice.

Reply:
For `PER_CLUSTER` method, currently we use `size_t big_enough     = 256ul * std::max<size_t>(index.pq_book_size(), index.pq_dim());` number of sample for each cluster. We could in theory just replace `256` with `max_train_points_per_pq_code` parameter. However, it would be quite confusing to users since the `big_enough` parameter depends on the max of `pq_book_size` and `pq_dim`, but `pq_dim` is not related to the number of pq_codes. If we name the knob as `max_train_points_per_pq_code`, the right way to calculate `big_enough` should be `size_t big_enough     = max_train_points_per_pq_code* index.pq_book_size();`, without `pq_dim` in the equation. I think we could either 

- Remove the `pq_dim` dependency when calculating `big_enough`. 
- Rename the knob to something more generic. Any suggestions on the possible names?

Reply:
I would keep the PER_CLUSTER formula the same (apart from introducing the new parameter instead of 256).

I think `max_train_points_per_pq_code` name is still ok: it conveys a meaning, even if it is not exact. Usually we would have `nlist` >> `pq_dim`, and in that case the name is corret.

Alternatively, you can consider `pq_train_points`, which is short, and the docstring can explain the details.

Reply:
I'll keep using `max_train_points_per_pq_code` and add comments in `PER_CLUSTER` function to explain